391 aperture scatterguard ophyd async

katesmith280 opened the pull request:
Fixes #391

### Instructions to reviewer on how to test:
1. Run tests

### Checks for reviewer
- [ ] Would the PR title make sense to a scientist on a set of release notes
- [ ] If a new device has been added does it follow the [standards](https://github.com/DiamondLightSource/dodal/wiki/Device-Standards)

DominicOram commented on src/dodal/devices/util/motor_utils.py:
```diff
@@ -1,6 +1,16 @@
     max_velocity: Component[EpicsSignalRO] = Component(EpicsSignalRO, ".VMAX")
+
+
+class ExtendedMotor(Motor):
```
I forgot that I had done some work to get these into `ophy-async` ages ago. I've revived it at https://github.com/bluesky/ophyd-async/pull/174. Let's not wait for that to be merged though, we can merge this in here and remove later

dperl-dls commented on src/dodal/devices/aperturescatterguard.py:
```diff
@@ -1,7 +1,6 @@
 from collections import namedtuple
 from dataclasses import dataclass
-from typing import List, Optional
 
```
you will want to `from bluesky.protocols import Movable`

dperl-dls commented on src/dodal/devices/aperturescatterguard.py:
```diff
@@ -85,7 +84,7 @@ class ApertureScatterguard(StandardReadable):
     def __init__(self, prefix: str = "", name: str = "") -> None:
```
`ApertureScatterguard` should inherit from both `StandardReadable` but also `Moveable`, which defines what the signature for `set()` should look like

dperl-dls commented on src/dodal/devices/aperturescatterguard.py:
```diff
@@ -116,37 +122,49 @@ def _get_motor_list(self):
-        return reduce(
-            operator.and_, [motor.set(pos) for motor, pos in zip(motors, positions)]
+    async def _set_raw_unsafe(self, positions: ApertureFiveDimensionalLocation):
+        """Accept the risks and move in an unsafe way. Collisions possible."""
```
love this docstring

DominicOram replied:
Agreed!

dperl-dls commented on src/dodal/devices/aperturescatterguard.py:
```diff
@@ -83,24 +80,33 @@ def as_list(self) -> List[SingleAperturePosition]:
         self.aperture_positions = positions
 
-    def set(self, pos: SingleAperturePosition) -> StatusBase:
+    def set(self, pos: SingleAperturePosition):
```
set does need to return an `AsyncStatus` in the end, so that the Bluesky RunEngine knows what to do with it

dperl-dls reviewed (changes requested):
Almost there, just one more thing really - I put some comments in the code to explain. We should make sure that the `set` method conforms to what bluesky expects by making the device `Movable` and not only `Readable`

This change should be really small but please add one more test which checks that we can use the device in a plan in the RunEngine. If you write this test first it should fail because the RunEngine expects `set()` to return a `Status` of some kind

It will look something roughly like:
```python
import bluesky.plan_stubs as bps
def test_ap_sg_in_runengine(aperture_in_medium_pos: ApertureScatterguard, RE: RunEngine):
    test_position = ... # I'll leave this to you
    RE(bps.abs_set(aperture_in_medium_pos, test_position, wait=True)
    # assert that positions afterwards are equal to specified in the test position
```

katesmith280 commented:
@dperl-dls thanks for all your help! 

I've merged the main branch into this one now and your test works great :-)

dperl-dls reviewed (approved):
Looks great! @DominicOram might want to take a quick look since I have had so much involvement in writing this, but as far as I can see this is good to merge. Let's squash the commits into one before merging it, you can do that directly from the github interface

DominicOram reviewed (dismissed):
Great work, thank you @katesmith280! Couple of comments in code. More importantly (and annoyingly) this has broken Hyperion. You can test this if you run the tests on Hyperion main against this branch. I think most of the failures seem to be around the renaming of `user_setpoint`/`user_readback` in `ophyd-async`. This should be fixed in https://github.com/bluesky/ophyd-async/pull/174. Maybe we wait on this to be merged first :/

katesmith280 commented on tests/devices/unit_tests/test_aperture_scatterguard.py:
```diff
@@ -176,105 +204,113 @@ def set_underlying_motors(
-    set_status.wait()
-    selected_aperture = aperture_in_medium_pos.read()
+    set_status = await aperture_in_medium_pos.set(aperture_positions.MEDIUM)
+    print(set_status)
```
Done :-D

katesmith280 commented on src/dodal/devices/aperturescatterguard.py:
```diff
@@ -116,37 +123,49 @@ def _get_motor_list(self):
+            self.TOLERANCE_STEPS * await self.aperture.y.motor_resolution.get_value()
+        )
+        # extendedepicsmotor class has tolerance fields added
+        # ophyd-async epics motor may need to do the same thing - epics motor
```
Comment is now removed

katesmith280 commented:
Thanks for the review @DominicOram! I agree waiting so it doesn't break Hyperion is a good idea :-)

katesmith280 commented:
@DominicOram tested now against hyperion and it passes - hooray! 

Should be good to merge if you think it's all good :-)

dperl-dls reviewed (dismissed):
Sorry, I got the wrong idea about what happened here. We don't need the `ExtendedMotor` at all anymore thanks to the recent change to the ophyd-async motor, they can all just be replaced with `Motor`, and the `motor_resolution` is now called `deadband`

katesmith280 commented:
@dperl-dls good catch - my bad!

The ExtendedMotor is now removed and replaced with Motor, motor_readback is now also set to deadband, tests updated and have been run against dodal and hyperion repos - all tests for aperture scatterguard pass :-)

dperl-dls reviewed (approved):
Sorry this has taken so long, thanks for all your hard work on this @katesmith280 !
